### PR TITLE
Fix sorting functionality. Add integration tests

### PR DESCRIPTION
**Pull request by egorhm**
Sorting functionality fixed:

- __sort and __sort_desc fixed in common.py
- integration test for sorting added

### CODE REVIEW

File: src/ggrc/services/common.py [hunk 1]
```diff
@@ -421,15 +422,14 @@ def _get_matching_types(self, model):
...
-    if hasattr(mapper.c, 'updated_at'):
-      columns.append(mapper.c.updated_at.label('updated_at'))
-    # columns.append(self._get_polymorphic_column(model))
+    columns = [
+        mapper.primary_key[0].label('id'),
+        self._get_type_select_column(model).label('type'),
+    ]
+    for column_name in ('context_id', 'updated_at'):
+      if hasattr(mapper.c, column_name):
+        columns.append(getattr(mapper.c, column_name).label(column_name))
```
Comment (alieh-rymasheuski) [inline src/ggrc/services/common.py]: Nicely reformatted. :+1:

File: src/ggrc/services/common.py [hunk 2]
```diff
@@ -681,12 +686,24 @@ class Resource(ModelView):
...
       Indicates that a model object was DELETEd and has been removed from the
       database. The sender in the signal will be the model class of the DELETEd
-      resource. The followin garguments will be sent along with the signal:
+      resource. The following arguments will be sent along with the signal:
 
         :obj: The model instance removed.
         :service: The instance of Resource handling the DELETE request.
       """,)
 
+  def _check_accept_header(func):  #pylint: disable=no-self-argument
```
Comment (alieh-rymasheuski) [inline src/ggrc/services/common.py]: What about making it a `@staticmethod` or something? `pylint: disable` doesn't seem quite good to me.

File: test/integration/ggrc/services/test_common.py [hunk 3]
```diff
@@ -84,33 +85,73 @@ def tearDown(self):
...
+    if page is not None:
+      params.append('__page=%s' % page)
+    if page_size is not None:
+      params.append('__page_size=%s' % page_size)
+    if sort is not None:
+      params.append('__sort=%s' % sort)
+    if sort_desc is not None:
+      params.append('__sort_desc=%s' % 'true' if sort_desc else 'false')
+
+    return ('?'.join([URL_MOCK_COLLECTION, '&'.join(params)])
```
Comment (alieh-rymasheuski) [inline test/integration/ggrc/services/test_common.py]: IMO,
```python
 '{url}?{params}'.format(url=URL_MOCK_COLLECTION, params='&'.join(params))
``` 
looks better than
```python
'?'.join([URL_MOCK_COLLECTION, '&'.join(params)])
```

File: test/integration/ggrc/services/test_common.py [hunk 4]
```diff
@@ -84,33 +85,73 @@ def tearDown(self):
...
-  def mock_url(self, resource=None):
+  @staticmethod
+  def mock_url(resource=None, page=None, page_size=None, sort=None,
+               sort_desc=None):
     if resource is not None:
       return URL_MOCK_RESOURCE.format(resource)
-    return URL_MOCK_COLLECTION
 
-  def mock_model(self, **kwarg):
-    if 'id' not in kwarg:
```
Comment (alieh-rymasheuski) [inline test/integration/ggrc/services/test_common.py]: The `id` is generated by SQLAlchemy, do I get it right?

File: test/integration/ggrc/services/test_common.py [hunk 5]
```diff
@@ -84,33 +85,73 @@ def tearDown(self):
...
+    collection = response.json['test_model_collection']
+    models = collection['test_model']
+    return models, collection
+
   def assertRequiredHeaders(self, response, headers=None):
-    if headers is None:
-        headers = {'Content-Type': 'application/json'}
+    default_headers = {'Content-Type': 'application/json'}
+    if headers is not None:
+      default_headers.update(headers)
```
Comment (alieh-rymasheuski) [inline test/integration/ggrc/services/test_common.py]: I think we want to update the `headers` with `default_headers`, not the other way round. It just seems illogical.

I'd propose the following:
```python
default_headers = {'Content-Type': 'application/json'}
headers = headers or {}
headers.update(default_headers)
```

File: test/integration/ggrc/services/test_common.py [hunk 6]
```diff
@@ -147,49 +187,144 @@ def test_missing_resource_get(self):
 
   def test_collection_get(self):
     """Test collection GET method from common.py"""
-    date1 = datetime(2013, 4, 17, 0, 0, 0, 0)
-    date2 = datetime(2013, 4, 20, 0, 0, 0, 0)
+    last_modified = datetime(2013, 4, 2, 0, 0, 0, 0)
```
Comment (alieh-rymasheuski) [inline test/integration/ggrc/services/test_common.py]: This value is quite fragile and depends on `mock_models` implementation. I'd do something like that:
```python
models = self.mock_models(count=2)
last_modified = max(model.updated_at for model in models)
models_json = [model.to_json() for model in models)]
```

Comment (alieh-rymasheuski) [inline test/integration/ggrc/services/test_common.py]: Or either we can change the signature of `mock_models`: let it be like `def mock_models(self, count: int=1) -> models, models_in_json`. Then we can write the code above as:
```python
models, models_json = self.mock_models(count=2)
last_modified = max(models.updated_at for model in models)
```

In cases we don't need the models or their json representation, we can do `models, _ = self.mock_models(...)` or `_, models_json = self.mock_models(...)`.

File: test/integration/ggrc/services/test_common.py [hunk 7]
```diff
@@ -147,49 +187,144 @@ def test_missing_resource_get(self):
...
+    self.assertEqual(30, paging['total'])
+
+  def test_collection_get_sorting_single_attr(self):
+    """Test collection GET method with sorting of a single attribute"""
+    self.mock_models(count=3)
+    response = self.client.get(self.mock_url(sort='updated_at'),
+                               headers=self.headers())
+    resp_models, _ = self.parse_response(response=response)
+
+    expected = ['2013-04-0%sT00:00:00' % (i + 1) for i in range(3)]
```
Comment (alieh-rymasheuski) [inline test/integration/ggrc/services/test_common.py]: Can we get rid of this list (that depends on the implementation of `mock_models`) like this?
```python
models_json = self.mock_models(count=3)
...  # send request, parse response
expected = sorted(models_json, key=lambda m: m['updated_at'])
self.assertListEqual(resp_models, expected)
```

Comment (alieh-rymasheuski) [inline test/integration/ggrc/services/test_common.py]: The same applies for `test_collection_get_sorting_single_attr_asc` and `test_collection_get_sorting_single_attr_desc` (with `sorted(..., reverse=True)` in the second case).

File: test/integration/ggrc/services/test_common.py [hunk 8]
```diff
@@ -147,49 +187,144 @@ def test_missing_resource_get(self):
...
+  def test_collection_get_sorting_multiple_attr(self):
+    """Test collection GET method with sorting of multiple attributes"""
+    # create models twice so we have same updated_at for some models
+    self.mock_models(count=3)
+    self.mock_models(count=3)
+
+    response = self.client.get(self.mock_url(sort='updated_at,id'))
+    resp_models, _ = self.parse_response(response=response)
+
+    for model, expected in zip(resp_models, (('2013-04-01T00:00:00', 1),
```
Comment (alieh-rymasheuski) [inline test/integration/ggrc/services/test_common.py]: Here we can generate the `expected` list like in my previous comments too.

**Comment by alieh-rymasheuski** [comment]
@egorhm, please see my comments. I'm concerned that your tests depend on the dates that are generated in the current `mock_models` implementation, while you can get the values you hard-coded from the mock models easily.

On the other hand, my proposal can possibly make the tests untraceable if they fail in the future.

File: src/ggrc/services/common.py (hunk 2 above)
Comment (egorhm) [inline src/ggrc/services/common.py]: Agree on this point. Let's move it to module global scope.

File: test/integration/ggrc/services/test_common.py (hunk 4 above)
Comment (egorhm) [inline test/integration/ggrc/services/test_common.py]: Correct.

File: test/integration/ggrc/services/test_common.py (hunk 6 above)
Comment (egorhm) [inline test/integration/ggrc/services/test_common.py]: It's good proposal, but not in tests. My point here is that this approach decrease test readability and the need to generate json representation where it's not necessary. It's can be rather confusion when we'll extend/change this tests in future.

File: test/integration/ggrc/services/test_common.py (hunk 8 above)
Comment (egorhm) [inline test/integration/ggrc/services/test_common.py]: The same as for previous comment. It's good proposal, but not in tests. My point here is that this approach decrease test readability and I'd like to see the order and representation set explicitly.

File: test/integration/ggrc/services/test_common.py (hunk 6 above)
Comment (alieh-rymasheuski) [inline test/integration/ggrc/services/test_common.py]: OK, if you think my proposal will decrease readability severely, let's keep your test implementation.

File: test/integration/ggrc/services/test_common.py (hunk 7 above)
Comment (egorhm) [inline test/integration/ggrc/services/test_common.py]: Please see my comment below.

**Comment by alieh-rymasheuski** [comment]
Approved.

**Pull request merged**